File: algorithm/Stack/minStack/minStack.py
#!/usr/bin/python
#
#

class MinStack:
  #initialize your data structure here.
  def __init__(self):
    self.stack1 = []
    self.stack2 = []
    self.min    = sys.maxint

  #@param x, an integer
  #@return nothing
  def push(self,x):
    self.stack1.append(x)
    if x <= self.min:
      self.stack2.append(x)
      self.min = x

  #@return nothing
  def pop(self):
    tmp = self.stack1.pop()
    if tmp == self.min:
      self.stack2.pop()
      if len(self.stack2) == 0:
        self.min = sys.maxint
      else:
        self.min = self.stack2[len(self.stack2)-1]

  #@return an integer
  def top(self):
    return self.stack1[len(self.stack1)-1]

  #@return an integer
  def getMin(self):
    return self.stack2[len(self.stack2)-1]

File: algorithm/Stack/minStack/MinStack.java
public class MinStack {

  Stack<Integer> s1 = new Stack<Integer>();
  Stack<Integer> s2 = new Stack<Integer>();

  public void push(int x) {
    s1.push(x);
    if (x <= min) {
      s2.push(x);
      min = x;
    }
  }

  public void pop() {
    int tmp = s1.pop();
    if (tmp == min) {
      s2.pop();
      if (!s2.isEmpty())
        min = s2.peek();
      else
        min = Integer.MAX_VALUE;
    }
  }

  public int top() {
    return s1.peek();
  }

}//end class MinStack
